### PR TITLE
List databases stream impl

### PR DESCRIPTION
**Pull request by rylev**
This add stream support for listing databases. This changes `list_databases` to always return a `Stream`. This might not be what we want since streams are not the most ergonomic of Rust constructs, but it does work. 

I co-authored this with @yoshuawuyts

### CODE REVIEW

**Review by MindFlavor — APPROVED** [review]
Great work! 
Just a question, shouldn't we keep the non-streaming function too? 
Streams are very cumbersome to work with.

**Comment by yoshuawuyts** [comment]
@MindFlavor Heya! - Ryan is out for a couple of days, but we paired on this together yesterday so I think I can answer this!

### Ergonomics of using `Stream` / `AsyncIterator`
The Rust async working group expects to add "async iteration syntax" to language. In terms of ergonomics that should allow for minimal differences between synchronous and asynchronous iteration. That would enable the following API change:
```diff
- let mut db_list = client.list_databases(cx, opts);
- while let Some(res) = db_list.next().await {
-     for db in res? {
+ for res.await? in client.list_databases(cx, opts) {
+   for db in res {
        println!("{:?}", db);
    }
}
```

It's hard to make any exact assurances or share timelines (these things take a while); but there seems to be consensus this is something we should enable - which means at that point it'll mostly be a matter of filling out the details (which I plan to be involved in). It's not there quite yet; but using streams in our external API today should set us up nicely for when language ergonomics catch up!

### Single response vs list of responses

Something we deliberated on during our call was whether to expose an additional API for "get a single response". We decided against it, since getting one item from `client.list_databases` corresponds to making a single HTTP request. Which means at least from an efficiency angle we couldn't find a clear reason for it.

Also having a streaming API clearly signals that there may be more results in different pages, allowing people to opt-in to seeing only a single page. But by default they're guided to iterating over all pages, which is probably a good thing.

And finally adding in the challenge of naming (`list_databases_single_page`? `list_databases_once`?) we figured it wouldn't be worth it to add an API just for this.

I hope that rationale makes sense!

**Comment by MindFlavor** [comment]
> Also having a streaming API clearly signals that there may be more results in different pages, allowing people to opt-in to seeing only a single page. But by default they're guided to iterating over all pages, which is probably a good thing.

Excellent point, you sold me on this one! While the others are valid points, I'm all in for explicitness in function signatures: the more information we convey with the function signature, the less we rely on documentation (which may or may not have been read beforehand). 
With the streaming signature we make sure the end user knows we *could* give more info. Nice 💘!
 
PS: As an aside, for this very reason I'm also against #338 but I'll elaborate more on that PR.

File: sdk/cosmos/tests/cosmos_database.rs [hunk 1]
```diff
@@ -12,9 +13,10 @@ async fn create_and_delete_database() {
     let client = setup::initialize().unwrap();
 
     // list existing databases and remember their number
-    let databases = client
-        .list_databases(Context::new(), ListDatabasesOptions::new())
+    let databases = Box::pin(client.list_databases(Context::new(), ListDatabasesOptions::new()))
+        .next()
         .await
+        .unwrap()
```
Comment (MindFlavor) [inline sdk/cosmos/tests/cosmos_database.rs:19]: This one fails with:

```
---- create_and_delete_database stdout ----
thread 'create_and_delete_database' panicked at 'called `Option::unwrap()` on a `None` value', sdk/cosmos/tests/cosmos_database.rs:19:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

I'll investigate further and get back to you, for the moment I'll revert the review.

**Review by MindFlavor — CHANGES_REQUESTED** [review]
Some tests are failing, I'll investigate further and get back to you ASAP.

File: sdk/cosmos/src/clients/cosmos_client.rs [hunk 2]
```diff
@@ -158,28 +163,80 @@ impl CosmosClient {
...
+                };
+
+                let response = r#try!(response);
+
+                let continuation_token = response
+                    .continuation_token
+                    .clone()
+                    .map(|ct| State::Continuation(ct))?;
+
+                Some((Ok(response), continuation_token))
```
Comment (MindFlavor) [inline sdk/cosmos/src/clients/cosmos_client.rs:237]:
```suggestion
                let next_state = response
                    .continuation_token
                    .clone() 
                    .map(|ct| State::Continuation(ct)) 
                    .unwrap_or_else(|| State::Done);           

                Some((Ok(response), next_state))
```

The above code incorrectly mapped the absence of `continuation_token` to the absence of data (which caused the bug). A simple `.unwrap_or_else(|| State::Done)` should be enough!

As an aside, I suggest renaming the binding from `continuation_token` to the more appropriate `next_state`.

**Review by MindFlavor — CHANGES_REQUESTED** [review]
I think I've pinpointed the bug. 
I'll try and run the tests again once you have addressed it!

(note: the E2E tests still fail because we haven't merged #339 yet).

**Review by yoshuawuyts — APPROVED** [review]

**Pull request merged**